### PR TITLE
feat(p3): TransformerPolicyNetwork BC-fit infrastructure (#281)

### LINKED ISSUE #281: diag(p3): TransformerPolicyNetwork BC-fit (Phase 1.5 escalation if #278+#279 fail)
## Hypothesis being tested

If both class-balanced BC (#278) and hidden_size sweep (#279) fail to recover the WORK-house discriminating sub-task, the architectural gap isn't pure capacity — it's *kind*: the MLP may not have the inductive biases needed to route `agent_identity → owned-house-index → houses[that_index] is burning?`. A Transformer's attention mechanism handles this kind of indexed-lookup naturally.

## Context

A `TransformerPolicyNetwork` already exists in `bucket_brigade/training/networks.py` (per PR #276 builder's note). This is a 1-line architecture swap.

## Protocol

Run BC-fit (class-balanced if #278 indicated it helps) with TransformerPolicyNetwork instead of MLP. Compare against the best MLP from #279.

## Verdict logic

| TransformerPolicyNetwork on house-on-WORK | Interpretation |
|---|---|
| **≥ 0.90** | Attention solved it. PPO's MLP was the wrong inductive bias. Next: full PPO with Transformer. |
| **No better than best MLP** | Architecture wasn't the gap. Specialist may use information that's not in the obs — fundamental task reformulation needed. |
| **Modest improvement** | Combined approach (Transformer + bigger hidden_size + class balancing) might work. File production training. |

## Cost

~1 hour. Same protocol as #272/#278/#279 with different network class.

## Dependency

**Blocked on #279.** Don't run if MLP+capacity already solves it.

## Acceptance criteria

- [ ] BC with Transformer policy net
- [ ] Compared against #279 best MLP
- [ ] Verdict applied; production-PPO follow-up filed if successful

## References

- PR #276 (#272 verdict)
- #278, #279 (Phase 1.5 sequence — must complete first)
- `bucket_brigade/training/networks.py::TransformerPolicyNetwork` (existing, ready to use)

---

### Status update: this issue is likely deprecated by #278's cross-finding

PR #278 (merged 2026-05-17) demonstrated that the **default MLP `PolicyNetwork(hidden_size=64)`** reaches BC `gap_closed = 0.934` at 40k pairs/agent × 30 epochs. This directly contradicts #272's `INDUCTIVE_BIAS_GAP` verdict and implies the MLP *does* have the inductive bias to route `agent_identity → owned-house-index → houses[that_index]`. The Transformer-escalation premise (capacity/architecture is the bottleneck) is no longer the leading hypothesis.

**Recommended scope** (in priority order):

1. **Defer until #279 resolves.** If class-balanced BC (#279) confirms #278's 0.934 result holds with the original (10k/10ep) budget after rebalancing, this issue should be closed as resolved (mark `loom:blocked` flag below stays until human decides).
2. **If #279 *and* the (now-likely-deprecated) #280 hidden_size sweep both come back negative on house-on-WORK accuracy**, then escalate to Transformer per the original protocol below.
3. **Orthogonal motivation**: even if not needed diagnostically, Transformer vs MLP on full PPO production is a legitimate comparison — but that's a separate issue, not this one.

### Dependencies

- [x] #278: BC-init + PPO continuation (MERGED — cross-finding documented)
- [ ] #279: Class-balanced BC-fit (open, curating) — **must complete before this issue runs**
- [ ] #280: hidden_size sweep (open, curating) — likely deprecated by #278's finding but listed for completeness

This issue cannot proceed until #279 returns a verdict. If #279 closes the capacity-vs-imbalance question affirmatively (MLP solves house-on-WORK with rebalancing), this issue should be closed without execution.

### Problem statement

If both class-balanced BC (#279) and hidden_size capacity (#280) fail to recover the WORK-house discriminating sub-task at adequate budget, the architectural gap is *kind*, not *size*: an MLP may lack the inductive biases to route `agent_identity → owned-house-index → houses[that_index] is burning?`. A Transformer's self-attention handles indexed lookup naturally. This is a Phase 1.5 escalation, gated on #279/#280 failure.

### Hypothesis under test

`BC(TransformerPolicyNetwork)` reaches house-on-WORK accuracy ≥ 0.90 on the same demonstrations that the MLP fails to fit, isolating attention-based indexed lookup as the missing inductive bias.

### Implementation guidance

`TransformerPolicyNetwork` already exists at `bucket_brigade/training/networks.py:324` with this signature:

```python
TransformerPolicyNetwork(
    obs_dim: int,
    action_dims: List[int],
    num_houses: int = None,        # auto-detected from obs_dim
    house_features: int = 4,        # fire_level, bucket_count, distance, threatened
    d_model: int = 256,             # transformer embedding dim
    nhead: int = 4,                 # attention heads
    num_layers: int = 3,            # transformer encoder layers
    dim_feedforward: int = 512,
    dropout: float = 0.1,
)
```

~350K params at defaults vs ~14K for `PolicyNetwork(hidden_size=64)`.

**Architectural fit (verify before running)**: confirm the obs layout fed to `bc_fit_only.py` (post-#204 identity-tail one-hot) matches the Transformer's assumed `[houses × 4 features + global]` layout. The Transformer auto-detects `num_houses` from `obs_dim`; if global features (identity tail, etc.) make `obs_dim % house_features != 0` the remainder is routed through `global_embed`. Sanity-check the parse: with `num_agents=2`, post-#204 obs ≈ `num_houses*4 + num_agents` so a 10-house env has `obs_dim = 42`, `num_houses = 10`, `global_dim = 2`. Good.

**Protocol** (mirror #279's, swap network class only):

1. Reuse `bc_fit_only.py`'s data path and verdict harness; add a `--network {mlp,transformer}` CLI flag (or write a `bc_fit_transformer.py` sibling — builder discretion).
2. Use whichever BC budget produced #279's best MLP result (likely 40k pairs / 30 epochs at minimum).
3. If #279 found class balancing helps, apply the same weighting here.
4. Compare house-on-WORK accuracy and `gap_closed` against #279's best MLP cell.
5. Same 3-seed protocol as the diagnostic ladder.

### Verdict logic

| Transformer house-on-WORK | Interpretation | Next step |
|---|---|---|
| ≥ 0.90 *and* clearly beats MLP | Attention solved indexed lookup. PPO's MLP was wrong inductive bias. | File PPO-with-Transformer production issue. |
| Within ~0.05 of best MLP | Architecture wasn't the gap. Specialist may use info not in obs — task reformulation needed. | File obs/reward redesign issue (env-side intervention per project thesis). |
| Modest improvement (0.05–0.15) | Combined approach (Transformer + balancing + budget) plausible but inconclusive. | Decide based on cost; default = file env-side issue. |

### Affected Files

- `experiments/p3_specialization/bc_fit_only.py:164` — current MLP instantiation; add `--network` flag here or fork into a transformer-specific module.
- `experiments/p3_specialization/bc_init.py:128` — sibling BC harness; same instantiation pattern.
- `bucket_brigade/training/networks.py:324` — `TransformerPolicyNetwork` (use as-is; no edits expected).
- `bucket_brigade/training/__init__.py:8` — already re-exports `TransformerPolicyNetwork`.
- `experiments/marl/train_gpu.py:280` — reference for an existing Transformer instantiation pattern (don't import from here; just match the constructor call).
- `experiments/p3_specialization/diagnostics/results/` — write results JSON alongside #278/#279 artifacts.

### Test Plan

- [ ] Sanity: parameter count for `TransformerPolicyNetwork(obs_dim=42, action_dims=[10,2,2])` matches docstring's ~350K (within 2×).
- [ ] Forward-pass shape check: instantiate net, run one batch through, assert action-head logits have shape `(B, 10)`, `(B, 2)`, `(B, 2)`.
- [ ] Reproducibility: 3 seeds; report mean ± stdev for `gap_closed` and per-head accuracies.
- [ ] Verdict cell applied per table above; results JSON committed to `experiments/p3_specialization/diagnostics/results/`.
- [ ] Cross-comparison: side-by-side house-on-WORK accuracy vs #279's best MLP cell.
- [ ] Cost gate: experiment runs in < 2 hours on `COMPUTE_HOST_PRIMARY`; abort and re-scope if blown.

### References

- PR #278 (cross-finding: MLP @ 64-hidden, 40k/30ep → `gap_closed = 0.934`)
- PR #276 / #272 (original `INDUCTIVE_BIAS_GAP` verdict, now contradicted)
- #279 (class-balanced BC; primary blocker)
- #280 (hidden_size sweep; secondary blocker)
- `bucket_brigade/training/networks.py:324` (`TransformerPolicyNetwork`)
- `bucket_brigade/training/networks.py:13` (`PolicyNetwork` baseline)

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Adds the **TransformerPolicyNetwork BC-fit infrastructure** for issue #281 as a Phase-1.5 escalation tool. Default behaviour is bit-identical to pre-#281 (MLP path).

**NB**: PR #278 (merged) found `PolicyNetwork(hidden_size=64)` reaches `gap_closed = 0.934` on `minimal_specialization` at 40k pairs / 30 epochs, which **contradicts the `INDUCTIVE_BIAS_GAP` premise** that motivated this escalation. Curator recommended close-without-execution if #279 confirms. **This PR ships the infrastructure but you may decide to skip execution.**

## Changes

- `experiments/p3_specialization/bc_fit_only.py` — `--architecture {mlp,transformer}` CLI flag + transformer config knobs (`d_model`, `nhead`, `num_layers`, `dim_feedforward`, `dropout`). Adds `_build_network()` helper. Default `architecture='mlp'` is bit-identical to pre-#281.
- `experiments/p3_specialization/bc_init.py` — same flag added to `bc_fit_one_agent()` and `main()`. Saved state dicts are not cross-architecture compatible.
- `experiments/p3_specialization/analyze_281.py` (new) — aggregates per-seed `bc_fit_only_result.json` files, computes mean ± stdev across seeds, and applies a 5-tier verdict ladder against PR #278's MLP baseline (`house_acc_on_work_subset ≈ 0.93`).
- `experiments/p3_specialization/run_issue281_transformer_bc.sh` (new) — 1-cell sweep driver (3 seeds, 40k pairs/agent × 30 epochs to match #278). **DO NOT RUN LOCALLY**; cost gate <2h on `COMPUTE_HOST_PRIMARY`.
- `tests/test_bc_fit_transformer.py` (new) — 7 tests.

## Obs-Layout Sanity Check

Confirmed: post-#204 the env produces `obs_dim=42` with `num_houses=10, global_dim=2` for `num_agents=2`. `TransformerPolicyNetwork` auto-detects this correctly (verified in `test_transformer_forward_shape_post_204_layout`). The transformer treats houses as per-house tokens and prepends the global identity tail as a `[CLS]` token.

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| Reuse existing `TransformerPolicyNetwork` | done | Imported from `bucket_brigade.training.networks`; no new network class. |
| `--architecture` flag on `bc_fit_only.py:164` | done | Added in `train_bc()` + CLI |
| `--architecture` flag on `bc_init.py:128` | done | Added in `bc_fit_one_agent()` + CLI |
| Obs-layout sanity for post-#204 (42, 10, 2) | done | `test_transformer_forward_shape_post_204_layout` |
| Sweep driver `run_issue281_transformer_bc.sh` | done | 3 seeds, NOT executed (per scope) |
| Analyzer `analyze_281.py` w/ verdict ladder | done | 5 verdict tiers; explicit ref to PR #278 baseline |
| Bit-identity test at default `--architecture mlp` | done | `test_train_bc_mlp_bit_identical_to_pre_281` |
| Transformer forward-pass shape test | done | `test_transformer_forward_shape_post_204_layout` |
| Local 1-2 epoch transformer smoke | done | Ran 2 epochs × 50 pairs end-to-end with real env |
| Cost gate <2h on `COMPUTE_HOST_PRIMARY` | done | Recorded in sweep script docstring |

## Test Plan

- `pytest tests/test_bc_fit_transformer.py tests/test_bc_init.py` — **11 passed** locally
- CLI smoke (mlp): `bc_fit_only.py --num-steps 13 --epochs 2` and `--architecture mlp` produce **identical** outputs (verified eval_loss = 3.3053 in both cases at seed 0)
- CLI smoke (transformer): `bc_fit_only.py --architecture transformer --transformer-d-model 64 --transformer-num-layers 1` runs end-to-end against the real env, writes valid JSON, analyzer parses it correctly
- Analyzer smoke: synthesized 2-seed result dir, analyzer aggregated and produced the expected `TRANSFORMER_WORSE` verdict given the smoke-test inputs

## Cost Note

This PR ships infrastructure only — no compute was consumed. Per task constraints, the sweep was **not executed**. The 1-cell sweep (3 seeds, 40k pairs / 30 epochs) is estimated <2h on `COMPUTE_HOST_PRIMARY` per the cost gate in the curator spec.

Closes #281

### CODE REVIEW

**Comment by rjwalters** [comment]
✅ **Approved!**

Phase-1.5 transformer escalation infrastructure looks solid. Default `--architecture mlp` preserves bit-identical pre-#281 behaviour and all 11 targeted tests pass locally.

## Verification

- All 6 CI checks pass (build-web, integration-test, lint-python, pre-commit, test-python, test-web).
- Merge state CLEAN. PR description uses `Closes #281`.
- `pytest tests/test_bc_fit_transformer.py tests/test_bc_init.py -v` → **11 passed in 4.11s** locally.
- Reuses `bucket_brigade.training.networks.TransformerPolicyNetwork` (line 324). No new network class introduced.
- `--architecture {mlp,transformer}` flag added in both `bc_fit_only.py` and `bc_init.py`; saved state dicts are correctly noted as non-cross-architecture-compatible.
- Obs-layout sanity at the post-#204 layout verified: `obs_dim=42` auto-splits to `num_houses=10, global_dim=2, house_features=4` (confirmed by direct construction).
- MLP bit-identity path: confirmed `train_bc(...)` without `architecture=` returns identical eval_loss / per-head accuracies to `architecture='mlp'` (test passes; same RNG path).
- Transformer forward-pass shape sanity: action logits `[(B,10), (B,2), (B,2)]` + value `(B,1)` on the 8×42 batch as expected.
- Param-count envelope `[300k, 3M]`: actual default config lands at **1,719,951** params on the 42-dim layout (vs MLP's 7,887). The widened envelope is correct.
- Sweep driver (`run_issue281_transformer_bc.sh`) is well-documented, shellcheck-clean, set -euo pipefail, and explicitly marked DO-NOT-RUN-LOCALLY with the <2hr cost-gate noted.
- Analyzer (`analyze_281.py`) cleanly aggregates per-seed JSONs, applies the 5-tier verdict ladder against PR #278's `house_acc_on_work_subset ≈ 0.93` reference, and emits both stdout + JSON output.

## Note: Pre-existing docstring inaccuracy in `networks.py`

`TransformerPolicyNetwork`'s docstring claims `~350000` params (line 357-359), but the default config (`d_model=256, nhead=4, num_layers=3, dim_ff=512`) actually produces ~1.72M params on the 42-dim obs layout. The builder correctly widened the param-count test to `[300K, 3M]` to accommodate reality. **This docstring is from PR #225, not this PR**, so it is not a blocker here — but a follow-up issue to fix the docstring (or right-size the default config) would be valuable. Not creating the issue here to avoid noise; flagging for the human reviewer's awareness.

## Reminder per Curator: may be obsolete given #278

The PR body and analyzer both transparently flag that **PR #278 already established `PolicyNetwork(hidden_size=64)` reaches `gap_closed = 0.934` on `minimal_specialization` at 40k pairs / 30 epochs**, which contradicts the `INDUCTIVE_BIAS_GAP` premise motivating #281. The Curator's recommendation to "close without execution if #279 confirms" still stands. This PR ships the infrastructure cleanly so the human can make that call without rework.

## Test Execution

Test plan from PR description:
1. `pytest tests/test_bc_fit_transformer.py tests/test_bc_init.py` → ✅ Executed: 11/11 pass locally in 4.11s.
2. CLI smoke (mlp): identical eval_loss across default / explicit `--architecture mlp` → ✅ Executed via `test_train_bc_mlp_bit_identical_to_pre_281`.
3. CLI smoke (transformer): end-to-end against real env → ⚠️ Skipped (requires CLI subprocess setup; covered by `test_train_bc_transformer_smoke` and `test_bc_fit_one_agent_transformer_smoke` which exercise the same code paths via direct Python calls).
4. Analyzer smoke (synth 2-seed dir): not directly executed in this review — accepted on inspection (logic is straightforward JSON aggregation + threshold check).

LGTM — ready for merge. Execution of the sweep itself is the human's call given #278's contradiction of the original premise.
